crates/runtime/src/lib.rs

```diff
@@ -142,6 +145,29 @@ impl<'a, T> GuestPtr<'a, [T]> {
         let base = GuestPtr::new(self.mem, self.offset_base());
         (0..self.len()).map(move |i| base.add(i))
     }
+
+    /// Safety note: The user must use the same `GuestBorrows` to check each
+    /// use of GuestType::as_raw that is alive simultaneously.
+    /// The resulting `*mut [T]` can be used as a `&mut [T]` as long as the
+    /// reference is dropped before any Wasm code is re-entered.
+    pub fn as_raw(&self, bc: &mut GuestBorrows) -> Result<*mut [T], GuestError> {
+        let ptr = self
+            .mem
+            .validate_size_align(self.pointer.0, T::guest_align(), self.pointer.1)?
+            as *mut T;
+
+        bc.borrow(Region {
+            start: self.pointer.0,
+            len: self.pointer.1 * T::guest_size(),
+        })?;
+
+        // SAFETY: iff there are no overlapping borrows (all uses of as_raw use this same
+        // GuestBorrows), its valid to construct a *mut [T]
```

Comment:
Technically this shouldn't even be `unsafe` actually. This should call [this method](https://doc.rust-lang.org/nightly/std/ptr/fn.slice_from_raw_parts_mut.html) ideally, but that's being stabilized in the next release. Otherwise there's a number of other safety guarantees which need to be upheld beyond overlapping borrows (like the guest/host memory layout, validity of the bytes, etc) which would need to be satisfied to start using this as `&mut [u8]`.

Reply:
I'll replace all these with the excellent docs in your other PR

Reply:
Merged master into this branch. Used your tests where possible.